rust/connlib/tunnel/src/peer.rs

```diff
@@ -537,104 +534,109 @@ struct TranslationState {
     }
 
     /// We define a [`TranslationState`] as used if we have seen outgoing traffic in the last 10s.
     fn is_used(&self, now: Instant) -> bool {
-        now.duration_since(self.last_outgoing) <= Duration::from_secs(10)
+        self.last_outgoing.is_some_and(|last_outgoing| {
+            now.duration_since(last_outgoing) <= Duration::from_secs(10)
+        })
     }
 
     fn is_expired(&self, now: Instant) -> bool {
         let ack_grace_period_expired =
-            now.duration_since(self.missing_responses_detected_at) >= Duration::from_secs(1);
+            self.missing_responses_detected_at
+                .is_some_and(|missing_responses_detected_at| {
+                    now.duration_since(missing_responses_detected_at) >= Duration::from_secs(1)
+                });
 
-        self.no_response_in_120s(now) && self.is_used(now) && ack_grace_period_expired
+        self.no_incoming_in_120s(now) && self.is_used(now) && ack_grace_period_expired
     }
 
-    fn no_response_in_120s(&self, now: Instant) -> bool {
+    fn no_incoming_in_120s(&self, now: Instant) -> bool {
         const CONNTRACK_UDP_STREAM_TIMEOUT: Duration = Duration::from_secs(120);
 
-        self.no_response_in_window(now - CONNTRACK_UDP_STREAM_TIMEOUT, now)
+        self.no_incoming_in_window(now - CONNTRACK_UDP_STREAM_TIMEOUT, now)
     }
 
-    fn no_response_in_window(&self, start: Instant, finish: Instant) -> bool {
-        let before_start = self.last_incoming < start;
-        let after_finish = self.last_incoming > finish;
+    fn no_incoming_in_window(&self, start: Instant, finish: Instant) -> bool {
+        let Some(last_incoming) = self.last_incoming else {
+            return true;
+        };
+
+        debug_assert!(start <= finish);
+
+        let before_start = last_incoming < start;
+        let after_finish = last_incoming > finish;
 
         before_start || after_finish
     }
 
     fn on_incoming_traffic(&mut self, now: Instant) {
-        self.last_incoming = now;
+        self.last_incoming = Some(now);
     }
 
-    fn this_packets_will_make_grace_period_start(&self, now: Instant) -> bool {
-        self.no_response_in_120s(now + Duration::from_secs(10))
-            != self.no_response_in_120s(self.last_outgoing + Duration::from_secs(10))
+    fn missing_multiple_responses(&self, now: Instant) -> bool {
+        /// The maximum time we expect a packet to be acknowledged in.
+        const MAX_ACK_DELAY: Duration = Duration::from_secs(110);
+
+        let no_incoming_before_last_packet = self
+            .last_outgoing
+            .map(|last_outgoing| {
+                self.no_incoming_in_window(last_outgoing - MAX_ACK_DELAY, last_outgoing)
+            })
+            .unwrap_or(true); // No outgoing packet means we also don't have a response!
+        let no_incoming_now = self.no_incoming_in_window(now - MAX_ACK_DELAY, now);
+
+        no_incoming_now && no_incoming_before_last_packet
```

Comment:
I think this is where the bug comes from, can you expand a bit on the logic behind this?

Before this change the function `this_packet_will_make_grace_period_start` meant that a given outgoing packet will trigger the `is_expired` due to the connection being used after that packet and more than 120 seconds happening.

Reply:
> I think this is where the bug comes from, can you expand a bit on the logic behind this?

I wanted to get rid of the `!=` comparison because I couldn't make sense of it and when I coded it into a `match (bool, bool)` with 4 branches, I noticed that some branches didn't matter for the tests to pass.

Reply:
> Before this change the function `this_packet_will_make_grace_period_start` meant that a given outgoing packet will trigger the `is_expired` due to the connection being used after that packet and more than 120 seconds happening.

My attempt was to refactor this so I get an understanding of what we are doing here but maybe I misunderstood? Feel free to push more commits!

Reply:
ah! the `!=` meant that the packet itself was the first to trigger it, let me try to give this another shot :)